design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should observe the underlying Kubernetes Service and expose whether there are sufficient healthy Endpoints.
+This information, analogous to Deployments observing and exposing Pod status, can be published as a `Ready` condition.
+
+Each `SupportService` CRD generates a unique upstream Envoy [Cluster][4], which will emit standard Envoy metrics.
+Note that the Envoy cluster name can be non-obvious, so exposing it in status may be helpful.
+
+If the `Service` refers to a Kubernetes `ExternalName`, Contour should program Envoy to send the traffic to the external destination.
+
+The `SupportService` CRD re-uses the `Service` type from the `projectcontour.io/v1` API.
+However, the setting following fields can generate a validation errors:
+
+- `Mirror` must not be set
+- `Weight` must not be set
+- `Protocol` may only be set to `h2`
+
+Note that the `Service` type does not include a field for a namespace name.
+This constrains `SupportService` resources to be in the same namespace as the `Service` they expose,
+ensuring that whoever creates the `SupportService` also has authority over the `Service` backing it.
+
+### HTTPProxy Changes
+
+```Go
+// SupportServiceReference names a SupportService resource.
+type SupportServiceReference struct {
+	// API version of the referent.
+	// +required
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,5,opt,name=apiVersion"`
+	// Kind of the referent.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +required
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+	// Namespace of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +required
+	Name string `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+}
+
+type AuthorizationPolicy struct {
+	// When true, this field disables client request authentication
+	// for the scope of the policy.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value pairs that are sent to the
+	// authentication server in the check request. If a context
+	// is provided at an enclosing scope, the entries are merged
+	// such that the inner scope overrides matching keys from the
+	// outer scope.
+	//
+	// +optional
+	Context map[string]string `json:"context,omitempty"`
+}
+
+type AuthorizationServer struct {
+	// +required
+	ServiceRef SupportServiceReference `json:"serviceRef"`
+
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
```

Comment:
Should we not allow `Disabled` at the root level? This essentially would allow users to configure auth, then disable it at the root and child routes would have to enable by setting `Disabled=false` on their route?

Reply:
I think that adding the special case makes it harder for users to reason about. The use case you present here seems like a reasonable one to me; or at least reasonably possible.